src/Reflection/ReflectionClass.php

```diff
@@ -415,17 +416,19 @@ private function getMethodsFromInterfaces(): array
      *
      * @return array<lowercase-string, ReflectionMethod> indexed by method name
      */
-    private function getMethodsIndexedByLowercasedName(): array
+    private function getMethodsIndexedByLowercasedName(AlreadyVisitedClasses $alreadyVisitedClasses): array
     {
+        $alreadyVisitedClasses->push($this->getName());
+
         if ($this->cachedMethods !== null) {
             return $this->cachedMethods;
         }
 
         $classMethods     = $this->getImmediateMethods();
         $className        = $this->getName();
-        $parentMethods    = $this->getParentMethods();
-        $traitsMethods    = $this->getMethodsFromTraits();
-        $interfaceMethods = $this->getMethodsFromInterfaces();
+        $parentMethods    = $this->getParentMethods(AlreadyVisitedClasses::createEmpty());
+        $traitsMethods    = $this->getMethodsFromTraits($alreadyVisitedClasses);
+        $interfaceMethods = $this->getMethodsFromInterfaces(AlreadyVisitedClasses::createEmpty());
```

Comment:
Wait... is it intentional to use empty instances for methods and interfaces, and not for traits? :thinking:

Reply:
Unfortunately yes. This is https://github.com/Roave/BetterReflection/pull/1256#discussion_r984661676

Reply:
I remember getting some false positive circular refs with enums I think. I can try finding out why, maybe it leads to some optimisations

Reply:
I found two popular cases where interfaces are visited multiple times at the moment via e.g. `getMethods()`:

1. an enum like

```php
enum StringEnum: string implements InterfaceForEnum
{
    case ONE
        = 'one';
    case TWO
        = 'two';
}
```
In that case apparently `UnitEnum` is automatically added as interface by PHP. Additionally `BackedEnum` is too. `getMethodsIndexedByLowercasedName` traverses then through the interface hierarchy, to query each interface for methods. It will look at `BackedEnum` and call `getMethodsIndexedByLowercasedName` again which get it's implemented interfaces. Since `BackendEnum` extends `UnitEnum` we end up with that thing again which leads to the circular reference.

2. A class Foo that implements interface A which extends interface B.

```php
interface B {}
interface A extends B {}
class Foo implements A {}
```

A very similar case where the interface hierarchy for Foo contains both interfaces already. And when traversing through them to get e.g. methods it looks at A and sees B again as implemented interface and that causes the circular reference.

I'm not sure yet how, but maybe this can be improved. CC @kukulich

Reply:
@herndlm I'm currently trying some optimization here: https://github.com/kukulich/BetterReflection/tree/optimization